sim/midas/targetutils/src/test/scala/RAMStyleHintSpec.scala

```diff
@@ -0,0 +1,99 @@
+  import midas.targetutils.xdc._
+  class MemoryModuleIO(addrBits: Int, dataBits: Int) extends Bundle {
+    val readAddress  = Input(UInt(addrBits.W))
+    val readEnable   = Input(Bool())
+    val readData     = Output(UInt(dataBits.W))
+
+    val writeAddress = Input(UInt(addrBits.W))
+    val writeData    = Input(UInt(dataBits.W))
+    val writeEnable  = Input(Bool())
+  }
+
+  abstract class BaseMemoryModule(annotator: MemBase[_] => Unit) extends Module {
+    // Arbitrarily selected.
+    val dataBits = 64
+    val addrBits = 16
+    val io = IO(new MemoryModuleIO(addrBits, dataBits))
+
+    def mem: MemBase[_]
+    annotator(mem)
+  }
+
+
+  class SyncReadMemModule(annotator: MemBase[_] => Unit) extends BaseMemoryModule(annotator) {
+    // Lazy so this is elaborated  before annotator executes
+    lazy val mem = SyncReadMem(1 << addrBits, UInt(dataBits.W))
+    io.readData := mem.read(io.readAddress, io.readEnable)
+    when(io.writeEnable) { mem(io.writeAddress) := io.writeData }
+  }
+
+  // This uses a combinational-read chisel memory, but could be infered as a BRAM
+  // due to the pipelining of the data read out. So the check annotator still accepts these.
+  class CombReadMemModule(annotator: MemBase[_] => Unit) extends BaseMemoryModule(annotator) {
+    lazy val mem = Mem(1 << addrBits, UInt(dataBits.W))
+    when(io.readEnable)  { io.readData          := RegNext(mem(io.readAddress)) }
+    when(io.writeEnable) { mem(io.writeAddress) := io.writeData }
+  }
+
+  class WrapperModule(annotator: MemBase[_] => Unit) extends Module {
+    val a = IO(new MemoryModuleIO(64, 16))
+    val b = IO(new MemoryModuleIO(64, 16))
+
+    val modA = Module(new SyncReadMemModule(_ => Nil))
+    val modB = Module(new SyncReadMemModule(_ => Nil))
+    modA.io <> a
+    modB.io <> b
+
+    annotator(modA.mem)
+    annotator(modB.mem)
+  }
+
+  def annotateMem(style: RAMStyle)(mem: MemBase[_]): Unit = RAMStyleHint(mem, style)
+
+  def elaborateAndCollectXDCAnnos(mod: =>Module) =
+    elaborate(mod)._2.collect { case a: XDCAnnotation => a }
+
+  behavior of "RAMStyleHint"
+
+  // Sanity check that stuff passes through elaboration. 
+  it should "correctly annotate a chisel3.SyncReadMem as BRAM" in {
+    val annos = elaborateAndCollectXDCAnnos(new SyncReadMemModule(annotateMem(RAMStyles.BRAM)))
+    assert(annos.size == 1)
+  }
+
+  it should "correctly annotate a chisel3.SyncReadMem as URAM" in {
+    val annos = elaborateAndCollectXDCAnnos(new SyncReadMemModule(annotateMem(RAMStyles.ULTRA)))
+    assert(annos.size == 1)
+  }
+
+  it should "correctly annotate a chisel3.Mem as URAM" in {
+    val annos = elaborateAndCollectXDCAnnos(new CombReadMemModule(annotateMem(RAMStyles.ULTRA)))
+    assert(annos.size == 1)
+  }
+
+  it should "correctly annotate a chisel3.Mem as BRAM" in {
+    val annos = elaborateAndCollectXDCAnnos(new CombReadMemModule(annotateMem(RAMStyles.BRAM)))
+    assert(annos.size == 1)
```

Comment:
Ummm.  You sure `annos.size == 1` is enough of an assertion here?  It's braver than I would be but you understand the innerworkings of chisel and annotation better than I.

Reply:
Yeah, this is not very precise. It didn't seem to me to be that useful to reconstitute the expected annotation but i'll just do that. 

If i push the machinery that actually generates the XDC into targetutils, i can test the end to end flow here. That might be more desireable in the long run.